README.md
# TeamTracker CLI

With `teamtracker` you are able to log your hours without leaving your IDE. Based on Puppeteer script parses your command, logs into your TeamTracker account and adds timelog entry.

You're able to have separate config file for each project you are working on. You can skip information provided there executing your command.

<p align="center">
  <img src="resources/usage.gif" width="1000" alt="Usage">
</p>

## Install

```bash
npm install --global teamtracker
```

or

```bash
yarn global add teamtracker
```

## Use

To log 1h 30m of Doing Something:
```bash
$ teamtracker timelog 1:30 Doing something
```

To change default task type use flag `--type`
```bash
$ teamtracker timelog 1:15 Release meeting --type Meeting
```

To define specific date use flag `--date`. 
(*by default you will log time for today*).
```bash
$ teamtracker timelog 0:25 Old task --date 2019-02-01
```

## Config file

In root of your project create file `.teamtrackerrc`.
In that file you can define flag values just to not pass them later from commandline 🤯

File may look like this:

```json
{
    "email": "michal.switon@venturedevs.com",
    "password": "yourSuperPassword",
    "project": "SuperSecretProject",
    "type": "Task"
}
```


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## 0.1.1 (2019-02-26)


### Features

* ability to log time. handling login errors ([1ffa0b0](https://github.com/croogie/teamtracker/commit/1ffa0b0))


src/cli.js
#!/usr/bin/env node

const chalk = require("chalk");
const meow = require("meow");
const keys = require("lodash/keys");

const timelog = require("./tasks/timelog");

const cli = meow(
  chalk`
Global flags:
{blue --debug -d}    Run in debug mode
{blue --verbose -v}  Run in verbose mode

{blue Config file}:
Place somewhere in your dir structure file: {blue .teamtrackerrc}. 
It can be in your project dir or even in your home directory.

Define content as follows:
{yellow.inverse \{
    "email": "you.name@venturedevs.com",
    "password": "yourPassword",
    "project": "LifeCents"
\}}

Then calling {blue timelog} command from project dir does not require to pass login credentials every time.

{red Remember!} to not add this file to your GIT repo! 

Available tasks:
{blue timelog} - ${timelog.help}
`,
  {
    booleanDefault: undefined,
    flags: {
      debug: {
        type: "boolean",
        default: false,
        alias: "d"
      },
      verbose: {
        type: "boolean",
        default: false,
        alias: "v"
      }
    }
  }
);

const tasks = {
  timelog
};

const taskName = cli.input[0];

if (tasks && tasks[taskName]) {
  const task = tasks[taskName];
  task(cli);
} else if (!taskName) {
  console.log(
    chalk`{red You have to define task!}
    
More information: {gray.inverse teamtracker --help}

Available tasks: {blue ${keys(tasks).join(", ")}}`
  );
} else {
  console.log(
    chalk`{red There's no task} {red.inverse ${taskName}}\n\nUse {gray --help} flag for more info.`
  );
}


src/tasks/timelog.js
const puppeteer = require("puppeteer");
const chalk = require("chalk");
const cosmiconfig = require("cosmiconfig");

const {
  isDebugging,
  getHours,
  getMinutes,
  composeMessage
} = require("../utils");

const logTime = async (
  page,
  description,
  hours,
  minutes,
  taskType,
  project,
  date = new Date().toJSON().slice(0, 10)
) => {
  process.stdout.write("👉 Logging time…");

  await page.waitForSelector("#id_when");
  await typeInInput(page, "#id_when", date);

  await page.waitForSelector(".timelog__project-field input.search");
  await typeInInput(page, ".timelog__project-field input.search", project);

  await page.waitForSelector(".timelog__type-field input.search");
  await typeInInput(page, ".timelog__type-field input.search", taskType);

  await page.waitForSelector("#id_hours");
  await typeInInput(page, "#id_hours", `${hours}`);

  await page.waitForSelector("#id_minutes");
  await typeInInput(page, "#id_minutes", `${minutes}`);

  await page.waitForSelector("#id_description");
  await typeInInput(page, "#id_description", description);

  const loading = page.waitForSelector("#timelog-submit-btn");
  await page.click("#timelog-submit-btn");

  await loading;

  process.stdout.write(
    chalk`\r👌 Successfully logged {red ${hours && hours + "h"}${minutes &&
      " " + minutes + "min"}} on project {red ${project}}!\r\n`
  );
};

export const loginUser = async (page, email, password) => {
  process.stdout.write(`👉 Logging as ${email}…`);

  await page.goto(
    "https://teamtrack.venturedevs.com/accounts/login/?next=/timelog/index/",
    { waitUntil: "networkidle2" }
  );

  await page.type("#id_login", email);
  await page.type("#id_password", password);

  const loginButtonSelector = ".login__button";
  await page.waitForSelector(loginButtonSelector);
  const awaiting = page.waitForNavigation({ waitUntil: "domcontentloaded" });

  await page.click(loginButtonSelector);
  await awaiting;

  try {
    const errorMessageSelector = ".ui.error.message > p";
    await page.waitForSelector(errorMessageSelector, { timeout: 200 });
    const error = await page.$(errorMessageSelector);

    process.stdout.write(`\r❌ Logging as ${email}\r\n`);
    const text = await page.evaluate(error => error.textContent, error);
    process.stdout.write(chalk`🙀 {red ${text}}\r\n`);

    return false;
  } catch (e) {
    process.stdout.write(`\r👌 Logging as ${email}\r\n`);

    return true;
  }
};

export const typeInInput = async (page, selector, text) => {
  const elementHandle = await page.$(selector);
  await elementHandle.click();
  await elementHandle.focus();
  await elementHandle.click({ clickCount: 3 });
  await elementHandle.press("Backspace");
  await elementHandle.type(text);
};

export const timelogTask = async ({ flags, input }) => {
  const [cmd, time, ...taskName] = input;

  const explorer = cosmiconfig("teamtracker");
  const configFound = await explorer.search();

  let flagsWithConfig = { ...flags };
  if (configFound) {
    flagsWithConfig = { ...configFound.config, ...flagsWithConfig };
  }

  if (!flagsWithConfig.email || !flagsWithConfig.password) {
    console.log(
      chalk`{red You have to provide} {red.inverse email} {red and} {red.inverse password} {red values!} \n\n💡 Use flags {red.inverse --email name@venturedevs.com} or {red.inverse --password yourPassword} or define in config file`
    );
    return;
  }

  if (!flagsWithConfig.project || !flagsWithConfig.type) {
    console.log(
      chalk`{red You have to provide} {red.inverse email} {red and} {red.inverse password} {red values!} \n\n💡 Use flags {red.inverse --project YourProjectName} or {red.inverse --type Meeting} or define in config file`
    );
    return;
  }

  let browser = await puppeteer.launch(isDebugging(flags.debug));
  let page = await browser.newPage();

  try {
    await loginUser(page, flagsWithConfig.email, flagsWithConfig.password);
    await logTime(
      page,
      composeMessage(...taskName),
      getHours(time),
      getMinutes(time),
      flagsWithConfig.type,
      flagsWithConfig.project,
      flagsWithConfig.date || undefined
    );
  } catch (error) {
    browser.close();
    process.exit(1);
  } finally {
    browser.close();
  }
};

timelogTask.help = chalk`Logs new activity to TeamTracker.

          {gray.inverse $ teamtracker timelog HH:MM Description [flags]}

          Required flags (if not defined in config):
            {gray --email} - User's email
            {gray --password} - User's password
            {gray --project} - Project where log time
            {gray --type} - Task type (possible values: {green Task}, {green Meeting}, {green Other})
            
          Optional flags:
            {gray --date} - Work date (format: {green 2019-02-20}). 
                     {gray If not defined will use today's date.}
          
          Examples: 
            {gray.inverse $ teamtracker timelog 1:30 Doing something}
            {gray.inverse $ teamtracker timelog 1:15 Release meeting --type Meeting}
            {gray.inverse $ teamtracker timelog 0:25 Old task --date 2019-02-01}`;

module.exports = timelogTask;
